File: Linked list/next_greater_node_in_linkedList.py
"""We are given a linked list with head as the first node.  
Let's number the nodes in the list: node_1, node_2, node_3, ... etc.

Each node may have a next larger value: for node_i, next_larger(node_i) 
is the node_j.val such that j > i, node_j.val > node_i.val, and j is 
the smallest possible choice.  If such a j does not exist, 
the next larger value is 0."""




# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def nextLargerNodes(self, head: ListNode) -> List[int]:
        if head is None:
            return 
        from collections import defaultdict
        d = {}
        d  = defaultdict(lambda:0,d)
        cur = head
        st = []
        while cur is not None:
            while st and st[-1].val < cur.val:
                d[st.pop()] = cur.val
            st.append(cur)
            cur = cur.next
        l = []
        while head:
            l.append(d[head])
            head = head.next
        return l

File: Linked list/NextGreaterNodeInLinkedList.cpp
//Problem Link:- https://leetcode.com/problems/next-greater-node-in-linked-list/

#include<bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        vector<int> result;
        ListNode* curr = head;
        while (curr != NULL){
            ListNode* temp = curr;
            bool flag = false;
            while (temp != NULL){
                if (temp->val > curr->val){
                    result.push_back(temp->val);
                    flag = true;
                    break;
                }
                temp = temp->next;
            }
            if (temp == NULL && flag == false){
                result.push_back(0);
            }
            curr = curr->next;
        }
        return result;
    }
};